### ISSUE TITLE
Reactions support

### ISSUE THREAD

**Issue by madamoniem** [flyerhq/flutter_chat_ui#394]
A similar issue was posted on August 2021, but there has been no update since.

Does the flutter_chat_ui package support reactions in any way? It would work by being a field withen a Message. This is an example of my current implementation of a TextMessage.

            if ((element.snapshot.value as Map)["type"] == "text") {
              _moreMessages.add(
                types.TextMessage(
                  **reactions: [],** <--- add this field to reactions
                  author: types.User(
                    id: (element.snapshot.value as Map)["author"]["id"],
                    firstName: (element.snapshot.value as Map)["author"]
                        ["firstName"],
                    lastName: (element.snapshot.value as Map)["author"]
                        ["lastName"],
                    imageUrl: (element.snapshot.value as Map)["author"]
                        ["imageUrl"],
                  ),
                  id: (element.snapshot.value as Map)["id"],
                  text: (element.snapshot.value as Map)["text"],
                  type: types.MessageType.text,
                  createdAt: (element.snapshot.value as Map)["createdAt"],
                ),
              );
            } 

It wouldn't need to be as complicated right now, just allow the user to pass in a list of emojis 👍 as text into the field. They can then be rendered using a Stack widget and positioned on the bottom left or right depending on the sender. A more advanced version would allow the user to pass in a map:

```
"reactions":[
   {
      "emoji":"👍",
      "reactors":[
         {
            "firstName",
            "lastName",
            "id"
         }
      ]
   }
]

```
This is how the basic reaction type would look like, where it only takes an array of emojis as text
![image](https://user-images.githubusercontent.com/98095976/216821754-4db6c461-0fe8-461a-ba06-36a00d11f14f.png)

This is how the more advanced one would look
![image](https://user-images.githubusercontent.com/98095976/216822036-a6599272-d732-458f-bd1f-3709c2b69dd3.png)

You would then have to add a "reactionBuilder," which would take in "p0" as other builders currently do, and the map of reactions that were passed on when building the Message, so the user can break it down, and customize it. You would allow the user to define what to do if an existing reaction is tapped on, how to style reactions that are already selected by the user, etc. To add a reaction, it would be preferrable to have a built in UI to handle reaction selection, but you can probably go with using the existing "onMessage_" functions in the Chat class.

Here is an existing package that implements the UI for reactions: https://pub.dev/packages/reaction_askany
You would preferably want the user to be able to select their own emoji, not from the few existing ones, however.

![image](https://user-images.githubusercontent.com/98095976/216823000-33af9f5b-c3d1-4c9c-bc1d-46fec072dadd.png)

**I am currently in dire need of this feature, and I have been trying to find ways to add it into my current codebase without using CustomMessage for every message, so if there are existing workarounds, such as adding custom styling on top of any Measge widget, or passing on reactions without having to redo all the built in Message types into CustomMessage, please let me know.**

I am currently using the flutter_chat_ui package heavily as a messaging interface, and reactions are a must have to make this package complete. I appreciate the work you've done in making a fantastic and easy to use package, and hope you look into adding something like this in the near future.

**Comment by zamirszn** [comment]
> A similar issue was posted on August 2021, but there has been no update since.
> 
> Does the flutter_chat_ui package support reactions in any way? It would work by being a field withen a Message. This is an example of my current implementation of a TextMessage.
> 
> ```
>         if ((element.snapshot.value as Map)["type"] == "text") {
>           _moreMessages.add(
>             types.TextMessage(
>               **reactions: [],** <--- add this field to reactions
>               author: types.User(
>                 id: (element.snapshot.value as Map)["author"]["id"],
>                 firstName: (element.snapshot.value as Map)["author"]
>                     ["firstName"],
>                 lastName: (element.snapshot.value as Map)["author"]
>                     ["lastName"],
>                 imageUrl: (element.snapshot.value as Map)["author"]
>                     ["imageUrl"],
>               ),
>               id: (element.snapshot.value as Map)["id"],
>               text: (element.snapshot.value as Map)["text"],
>               type: types.MessageType.text,
>               createdAt: (element.snapshot.value as Map)["createdAt"],
>             ),
>           );
>         } 
> ```
> 
> It wouldn't need to be as complicated right now, just allow the user to pass in a list of emojis 👍 as text into the field. They can then be rendered using a Stack widget and positioned on the bottom left or right depending on the sender. A more advanced version would allow the user to pass in a map:
> 
> ```
> "reactions":[
>    {
>       "emoji":"👍",
>       "reactors":[
>          {
>             "firstName",
>             "lastName",
>             "id"
>          }
>       ]
>    }
> ]
> ```
> 
> This is how the basic reaction type would look like, where it only takes an array of emojis as text ![image](https://user-images.githubusercontent.com/98095976/216821754-4db6c461-0fe8-461a-ba06-36a00d11f14f.png)
> 
> This is how the more advanced one would look ![image](https://user-images.githubusercontent.com/98095976/216822036-a6599272-d732-458f-bd1f-3709c2b69dd3.png)
> 
> You would then have to add a "reactionBuilder," which would take in "p0" as other builders currently do, and the map of reactions that were passed on when building the Message, so the user can break it down, and customize it. You would allow the user to define what to do if an existing reaction is tapped on, how to style reactions that are already selected by the user, etc. To add a reaction, it would be preferrable to have a built in UI to handle reaction selection, but you can probably go with using the existing "onMessage_" functions in the Chat class.
> 
> Here is an existing package that implements the UI for reactions: https://pub.dev/packages/reaction_askany You would preferably want the user to be able to select their own emoji, not from the few existing ones, however.
> 
> ![image](https://user-images.githubusercontent.com/98095976/216823000-33af9f5b-c3d1-4c9c-bc1d-46fec072dadd.png)
> 
> **I am currently in dire need of this feature, and I have been trying to find ways to add it into my current codebase without using CustomMessage for every message, so if there are existing workarounds, such as adding custom styling on top of any Measge widget, or passing on reactions without having to redo all the built in Message types into CustomMessage, please let me know.**
> 
> I am currently using the flutter_chat_ui package heavily as a messaging interface, and reactions are a must have to make this package complete. I appreciate the work you've done in making a fantastic and easy to use package, and hope you look into adding something like this in the near future.

Hey , hello any update regarding this ?

**Comment by luizwalber** [comment]
i would love to see this feature too

**Comment by fraboniface** [comment]
Same, this is literraly the first thing my users complain about in my entire app.

**Comment by zamirszn** [comment]
> Same, this is literraly the first thing my users complain about in my entire app. 

You have a chat app ??

**Comment by fraboniface** [comment]
> You have a chat app ??
I'm developing an app for which chat is an important functionality among others. It's not released yet, I was talking about beta testers.

**Comment by madamoniem** [comment]
I just stopped using the package entirely and made my own chat system with reactions, audio messages, camera and multiple image uploads, etc with way less code and the most accurate control over everything, Id recommend you to just face it and stop using the package, esp since reactions will never be implemented-

**Comment by demchenkoalex** [comment]
exactly what @madamoniem said, there is nothing better than something you build yourself. Having maintaining this for years now myself it is not easy to keep up. I started this when I had lots of energy and free time, and now my back hurts when I sit too much 😆 

Having said that I finally working on a v2 complete package re-design to make it more performant and extendable. I did not add reactions in v1 just because code here became so bad it hurts just looking at it, and performance is awful. So I am building something new that will solve this.

For now, there is nothing better than your own solution or feel free to use other packages :) I will close discussion here cause it will not get anywhere until v2 is complete.

Thanks everyone for understanding.
